Libraries/DirectMLX.h

```diff
@@ -4291,6 +4291,191 @@ namespace dml
+        }
+
+        TensorDesc queryTensor = query ? query->Impl()->GetOutputDesc() : TensorDesc();
+        TensorDesc keyTensor = key ? key->Impl()->GetOutputDesc() : TensorDesc();
+        TensorDesc valueTensor = value ? value->Impl()->GetOutputDesc() : TensorDesc();
+        TensorDesc stackedQueryKeyTensor = stackedQueryKey ? stackedQueryKey->Impl()->GetOutputDesc() : TensorDesc();
+        TensorDesc stackedKeyValueTensor = stackedKeyValue ? stackedKeyValue->Impl()->GetOutputDesc() : TensorDesc();
+        TensorDesc stackedQueryKeyValueTensor = stackedQueryKeyValue ? stackedQueryKeyValue->Impl()->GetOutputDesc() : TensorDesc();
+        TensorDesc biasTensor = bias ? bias->Impl()->GetOutputDesc() : TensorDesc();
+        TensorDesc maskTensor = mask ? mask->Impl()->GetOutputDesc() : TensorDesc();
+        TensorDesc relativePositionBiasTensor = relativePositionBias ? relativePositionBias->Impl()->GetOutputDesc() : TensorDesc();
+        TensorDesc pastKeyTensor = pastKey ? pastKey->Impl()->GetOutputDesc() : TensorDesc();
+        TensorDesc pastValueTensor = pastValue ? pastValue->Impl()->GetOutputDesc() : TensorDesc();
+        TensorDesc pastSequenceLengthsTensor = pastSequenceLengths ? pastSequenceLengths->Impl()->GetOutputDesc() : TensorDesc();
+
+        uint32_t batchSize;
+        uint32_t sequenceLength;
+        uint32_t headSize;
+        uint32_t valueHeadSize;
+        DML_TENSOR_DATA_TYPE dataType;
+
+        if (query)
+        {
+            assert(queryTensor.sizes.size() >= 3);
+            batchSize = queryTensor.sizes[queryTensor.sizes.size() - 3];
+            sequenceLength = queryTensor.sizes[queryTensor.sizes.size() - 2];
+            headSize = queryTensor.sizes[queryTensor.sizes.size() - 1] / queryHeadCount;
+            dataType = queryTensor.dataType;
+
+            if (value)
+            {
+                assert(valueTensor.sizes.size() >= 3);
+                valueHeadSize = valueTensor.sizes[valueTensor.sizes.size() - 1] / keyValueHeadCount;
+            }
+            else if (stackedKeyValue)
+            {
+                assert(valueTensor.sizes.size() >= 3);
+                valueHeadSize = valueTensor.sizes[valueTensor.sizes.size() - 1];
+            }
+        }
+        else if (stackedQueryKey)
+        {
+            assert(stackedQueryKeyTensor.sizes.size() >= 5);
+            batchSize = stackedQueryKeyTensor.sizes[stackedQueryKeyTensor.sizes.size() - 5];
+            sequenceLength = stackedQueryKeyTensor.sizes[stackedQueryKeyTensor.sizes.size() - 4];
+            headSize = stackedQueryKeyTensor.sizes[stackedQueryKeyTensor.sizes.size() - 1];
+            dataType = stackedQueryKeyTensor.dataType;
+        }
+        else
+        {
+            assert(stackedQueryKeyValue);
+            assert(stackedQueryKeyValueTensor.sizes.size() >= 5);
+            batchSize = stackedQueryKeyValueTensor.sizes[stackedQueryKeyValueTensor.sizes.size() - 5];
```

Comment:
`stackedQueryKeyValueTensor.sizes.size() - 5`

I worry about callers using DMLX and directly populating tensors from some model description, and then us accessing invalid negative indices because the tensor size is too small, especially if that model description comes from external data that is not completely under the program's control. We could say that it's the responsibility of the caller to validate all these sizes up-front before calling DMLX, but even DML validates tensor sizes before accessing any potentially invalid indices. Can we strengthen these mere asserts which only happen in debug builds to an `std::invalid_argument` instead?

```
#if __cpp_exceptions
    #if DMLX_USE_WIL
        #define DMLX_THROW_IF_FAILED(_hr) THROW_IF_FAILED(_hr)
        #define DMLX_THROW(_hr) THROW_HR(_hr)
        #define DMLX_THROW_IF_NOT(condition, exceptionType) if (!(condition)) { throw exceptionType; }
    #else
        #define DMLX_THROW_IF_FAILED(_hr) if (FAILED(_hr)) { throw std::runtime_error(#_hr); }
        #define DMLX_THROW(_hr) throw std::runtime_error(#_hr); 
        #define DMLX_THROW_IF_NOT(condition, exceptionType) if (!(condition)) { throw exceptionType; }
    #endif
#else
    #define DMLX_THROW_IF_FAILED(_hr) if (FAILED(_hr)) { std::abort(); }
    #define DMLX_THROW(_hr) { std::abort(); } 
    #define DMLX_THROW_IF_NOT(condition, exceptionType) { std::abort(); } 
#endif
```

```
            DMLX_THROW_IF_NOT(stackedQueryKeyValueTensor.sizes.size() >= 5);
```